File: web/src/lib/metadata_model/2d_fields.ts
import MetadataModel from '.'

/**
 * Extracts fields that will can used for working with data in 2D array form like in excel or csv.
 *
 * Recommended order to Extract fields:
 * * {@linkcode Extract2DFields.Extract}
 * * {@linkcode Extract2DFields.Reposition}
 * * {@linkcode Extract2DFields.RemoveSkipped}
 * * {@linkcode Extract2DFields.Fields} to get currrent state of fields.
 */
export class Extract2DFields {
	private _fields: (MetadataModel.IMetadataModel | any)[] = []
	private _skipIfFGDisabled: boolean = true
	private _skipIfDataExtraction: boolean = true
	private _removePrimaryKey: boolean = true
	private _metadatamodel: any
	private _repositionFields: MetadataModel.RepositionFields = {}

	/**
	 *
	 * @throws {MetadataModel.Error} if {@linkcode metadatamodel} is not valid
	 *
	 * @param metadatamodel A valid metadata-model of type object (not array). Expected to presented as if converted from JSON.
	 * @param skipIfFGDisabled Do not include field group if property {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE}($FG_VIEW_DISABLE) is true. Default true.
	 * @param skipIfDataExtraction Do not include field group if property {@linkcode MetadataModel.FgProperties.FIELD_GROUP_SKIP_DATA_EXTRACTION}($FG_SKIP_DATA_EXTRACTION) is true. Default true.
	 * @param removePrimaryKey Remove field if {@linkcode skipIfFGDisabled} or {@linkcode skipIfDataExtraction} even if {@linkcode MetadataModel.FgProperties.FIELD_GROUP_IS_PRIMARY_KEY} field property is true.
	 */
	constructor(metadatamodel: any, skipIfFGDisabled: boolean = true, skipIfDataExtraction: boolean = true, removePrimaryKey: boolean = true) {
		if (!MetadataModel.isGroupFieldsValid(metadatamodel)) {
			throw [Extract2DFields.name, 'argument metadatamodel is not an object']
		}

		this._metadatamodel = metadatamodel
		this._skipIfFGDisabled = skipIfFGDisabled
		this._skipIfDataExtraction = skipIfDataExtraction
		this._removePrimaryKey = removePrimaryKey
	}

	/**
	 * Return information about fields that need to be repositioned based on {@linkcode MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION} when {@linkcode Extract2DFields.Extract} is called.
	 */
	get RepositionFields() {
		return this._repositionFields
	}

	/**
	 * Return {@linkcode this._fields} in its current state.
	 */
	get Fields() {
		return this._fields
	}

	/**
	 * Extracts fields found in {@linkcode mmGroup} and places them in {@linkcode Extract2DFields._fields}.
	 *
	 * @throws {MetadataModel.Error} if {@linkcode Extract2DFields._extract} throws an error.
	 *
	 * Will not reposition fields based on {@linkcode MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION} field/group property. Call {@linkcode Extract2DFields.Reposition} after extraction.
	 *
	 * Will not remove any fields if {@linkcode MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION} or {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE} is true. Call {@linkcode Extract2DFields.RemoveSkipped} after extraction.
	 *
	 */
	Extract() {
		try {
			this._extract(this._metadatamodel)
		} catch (e) {
			throw e
		}
	}

	/**
	 * Recursive function that extracts fields found in {@linkcode mmGroup} and places them in {@linkcode Extract2DFields._fields}.
	 *
	 * Will not remove any fields if {@linkcode MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION} or {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE} is true.
	 *
	 * @throws {MetadataModel.Error} if {@linkcode mmGroup} is not valid.
	 *
	 * @param mmGroup Current metadata model Group to extract fields from.
	 * @param mmGroupSkipDataExtraction Add {@linkcode MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION} property to child fields and groups of {@linkcode mmGroup} if true.
	 * @param mmGroupViewDisable Add {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE} property to child fields and groups of {@linkcode mmGroup} if true.
	 *
	 * mm - Alias for metadata model.
	 *
	 * fg - Alias for field group.
	 */
	private _extract(mmGroup: any, mmGroupSkipDataExtraction: boolean = false, mmGroupViewDisable: boolean = false) {
		if (typeof mmGroup !== 'object' || Array.isArray(mmGroup) || mmGroup === null) {
			throw [this._extract.name, 'currentMetadataModelGroup is not an object']
		}

		const mmGroupFields = mmGroup[MetadataModel.FgProperties.GROUP_FIELDS][0]
		if (!MetadataModel.isGroupFieldsValid(mmGroupFields)) {
			throw [this._extract.name, `mmGroup${MetadataModel.FgProperties.GROUP_FIELDS}[0] is not an object`, structuredClone(mmGroupFields)]
		}

		const mmGroupReadOrderOfFields = mmGroup[MetadataModel.FgProperties.GROUP_READ_ORDER_OF_FIELDS]
		if (!MetadataModel.isGroupReadOrderOfFieldsValid(mmGroupReadOrderOfFields)) {
			throw [this._extract.name, `mmGroup${MetadataModel.FgProperties.GROUP_READ_ORDER_OF_FIELDS} is not an array`, structuredClone(mmGroupReadOrderOfFields)]
		}

		for (const fgKey of mmGroupReadOrderOfFields) {
			if (!MetadataModel.isGroupFieldsValid(mmGroupFields[fgKey])) {
				throw [this._extract.name, `mmGroupFields[${fgKey}] is not an object`, structuredClone(mmGroupFields[fgKey])]
			}

			if (Array.isArray(mmGroupFields[fgKey][MetadataModel.FgProperties.GROUP_FIELDS])) {
				const skipDataExtraction = (this._skipIfDataExtraction && mmGroupFields[fgKey][MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION]) || mmGroupSkipDataExtraction
				const viewDisable = (this._skipIfFGDisabled && mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE]) || mmGroupViewDisable

				if (mmGroupFields[fgKey][MetadataModel.FgProperties.GROUP_EXTRACT_AS_SINGLE_FIELD]) {
					this._fields.push(mmGroupFields[fgKey])
					continue
				}

				if (
					MetadataModel.GroupCanBeProcessedAs2D(mmGroupFields[fgKey]) &&
					mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS] &&
					!Number.isNaN(mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS]) &&
					mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS] > 0
				) {
					const maxIndexOfValuesInObject = Number(mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS]) - 1
					for (let columnIndex = 0; columnIndex <= maxIndexOfValuesInObject; columnIndex++) {
						for (let value of Object.values(mmGroupFields[fgKey][MetadataModel.FgProperties.GROUP_FIELDS][0])) {
							let newField: any = structuredClone(value)

							if (skipDataExtraction) {
								newField[MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION] = true
							}

							if (viewDisable) {
								newField[MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE] = true
							}

							newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] = columnIndex
							if (newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT]) {
								newField[MetadataModel.FgProperties.FIELD_GROUP_NAME] = (newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT] as string).replace(MetadataModel.ARRAY_PATH_REGEX_SEARCH, `${columnIndex + 1}`)
							} else {
								newField[MetadataModel.FgProperties.FIELD_GROUP_NAME] = `${newField[MetadataModel.FgProperties.FIELD_GROUP_NAME]} ${columnIndex + 1}`
							}

							this._fields.push(newField)
						}
					}
					continue
				}

				this._extract(mmGroupFields[fgKey], skipDataExtraction, viewDisable)
				continue
			}

			if (
				mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS] &&
				!Number.isNaN(mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS]) &&
				mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS] > 0
			) {
				const maxNoOfValuesInSeparateColumns = Number(mmGroupFields[fgKey][MetadataModel.FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS])

				for (let columnIndex = 0; columnIndex < maxNoOfValuesInSeparateColumns; columnIndex++) {
					let newField: any = structuredClone(mmGroupFields[fgKey])

					if (mmGroupSkipDataExtraction) {
						newField[MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION] = true
					}

					if (mmGroupViewDisable) {
						newField[MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE] = true
					}

					newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] = columnIndex
					if (newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT]) {
						newField[MetadataModel.FgProperties.FIELD_GROUP_NAME] = (newField[MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT] as string).replace(MetadataModel.ARRAY_PATH_REGEX_SEARCH, `${columnIndex + 1}`)
					} else {
						newField[MetadataModel.FgProperties.FIELD_GROUP_NAME] = `${newField[MetadataModel.FgProperties.FIELD_GROUP_NAME]} ${columnIndex + 1}`
					}

					this._fields.push(newField)
				}

				continue
			}

			let newField: any = structuredClone(mmGroupFields[fgKey])

			if (
				typeof newField[MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION] === 'object' &&
				!Array.isArray(newField[MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION]) &&
				typeof newField[MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION][MetadataModel.Field2dPositionProperties.FIELD_GROUP_KEY] === 'string'
			) {
				this._repositionFields[this._fields.length] = newField[MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION]
			}

			if (mmGroupSkipDataExtraction) {
				newField[MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION] = true
			}

			if (mmGroupViewDisable) {
				newField[MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE] = true
			}

			this._fields.push(newField)
		}
	}

	/**
	 * Calls {@linkcode Reposition2DFields}.
	 */
	Reposition() {
		this._fields = Reposition2DFields({ fields: this._fields, reposition: this._repositionFields })
	}

	/**
	 * Calls {@linkcode Reposition2DFields}.
	 */
	get FieldsRepositioned() {
		return Reposition2DFields({ fields: this._fields, reposition: this._repositionFields })
	}

	/**
	 * Calls {@linkcode RemoveSkipped2DFields}.
	 */
	RemoveSkipped() {
		this._fields = RemoveSkipped2DFields(this._fields, this._skipIfFGDisabled, this._skipIfDataExtraction, this._removePrimaryKey)
	}

	/**
	 * Calls {@linkcode RemoveSkipped2DFields}.
	 */
	get FieldsWithSkippedRemoved() {
		return RemoveSkipped2DFields(this._fields, this._skipIfFGDisabled, this._skipIfDataExtraction, this._removePrimaryKey)
	}
}

/**
 * Reorder columns of each row in data to match order in targetFields.
 */
export class Reorder2DFields {
	private _sourceToTargetReadOrderOfFields: number[]
	private _targetToSourceReadOrderOfFields: number[]

	get SourceToTargetReadOrderOfFields() {
		return this._sourceToTargetReadOrderOfFields
	}

	get TargetToSourceReadOrderOfFields() {
		return this._targetToSourceReadOrderOfFields
	}

	/**
	 *
	 * @param sourceFields Current order of columns in each row of data.
	 * @param targetFields Target order of columns in each row of data.
	 *
	 * @throws {MetadataModel.Error} if:
	 * * length of {@linkcode data} (each row), {@linkcode sourceFields} and {@linkcode targetFields} do not match.
	 * * field in {@linkcode sourceFields} is missing in {@linkcode targetFields}.
	 * * fields in {@linkcode sourceFields} and {@linkcode targetFields} are not valid.
	 *
	 */
	constructor(sourceFields: any[], targetFields: any[]) {
		if (sourceFields.length !== targetFields.length) {
			throw [Reorder2DFields.name, `length of sourceFields(${sourceFields.length}) and targetFields(${targetFields.length}) are not equal`, structuredClone(sourceFields), structuredClone(targetFields)]
		}

		this._sourceToTargetReadOrderOfFields = []
		this._targetToSourceReadOrderOfFields = new Array(sourceFields.length)

		for (let tfIndex = 0; tfIndex < targetFields.length; tfIndex++) {
			let sourceIndex = -1

			for (let sfIndex = 0; sfIndex < sourceFields.length; sfIndex++) {
				if (typeof sourceFields[sfIndex][MetadataModel.FgProperties.FIELD_GROUP_KEY] === 'string' && sourceFields[sfIndex][MetadataModel.FgProperties.FIELD_GROUP_KEY] === targetFields[tfIndex][MetadataModel.FgProperties.FIELD_GROUP_KEY]) {
					if (
						typeof sourceFields[sfIndex][MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] === 'number' &&
						sourceFields[sfIndex][MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] !== targetFields[tfIndex][MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX]
					) {
						continue
					}

					this._sourceToTargetReadOrderOfFields.push(sfIndex)
					sourceIndex = sfIndex

					break
				}
			}

			if (sourceIndex < 0) {
				throw [Reorder2DFields.name, 'targetField not found in sourceField', tfIndex, structuredClone(targetFields[tfIndex])]
			}

			this._targetToSourceReadOrderOfFields[sourceIndex] = tfIndex
		}
	}

	/**
	 * Reorder columns in each row in {@linkcode data}. Will modify order of columns in rows of original {@linkcode data} so deep copy may needed if original order is required.
	 *
	 * @throws {MetadataModel.Error} if length of each row of {@linkcode data} does not match {@linkcode Reorder2DFields._sourceToTargetReadOrderOfFields}
	 *
	 * @param data Rows of data to reorder.
	 */
	Reorder(data: any[][]) {
		for (let dIndex = 0; dIndex < data.length; dIndex++) {
			if (data[dIndex].length !== this._sourceToTargetReadOrderOfFields.length) {
				throw [this.Reorder.name, `length of datum at index ${dIndex} and targetFields are not equal`, structuredClone(data[dIndex])]
			}

			let targetDatum = new Array(this._sourceToTargetReadOrderOfFields.length)

			for (let fIndex = 0; fIndex < this._sourceToTargetReadOrderOfFields.length; fIndex++) {
				targetDatum[fIndex] = structuredClone(data[dIndex][this._sourceToTargetReadOrderOfFields[fIndex]])
			}

			data[dIndex] = targetDatum
		}
	}
}

/**
 * Remove 2D fields if {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE} or {@linkcode MetadataModel.FgProperties.FIELD_GROUP_SKIP_DATA_EXTRACTION} is true.
 * @param fields
 * @param skipIfFGDisabled Do not include field group if property {@linkcode MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE}($FG_VIEW_DISABLE) is true. Default false.
 * @param skipIfDataExtraction Do not include field group if property {@linkcode MetadataModel.FgProperties.FIELD_GROUP_SKIP_DATA_EXTRACTION}($FG_SKIP_DATA_EXTRACTION) is true. Default false.
 * @param removePrimaryKey Remove field even if it is a primary key.
 * @returns
 */
export function RemoveSkipped2DFields(fields: any[], skipIfFGDisabled: boolean = false, skipIfDataExtraction: boolean = false, removePrimaryKey: boolean = false) {
	return fields.filter((field) => {
		if (!removePrimaryKey) {
			if (field[MetadataModel.FgProperties.FIELD_GROUP_IS_PRIMARY_KEY]) {
				return true
			}
		}
		if (skipIfDataExtraction && field[MetadataModel.FgProperties.DATABASE_SKIP_DATA_EXTRACTION]) {
			return false
		}

		if (skipIfFGDisabled && field[MetadataModel.FgProperties.FIELD_GROUP_VIEW_DISABLE]) {
			return false
		}

		return true
	})
}

/**
 * Rearrange 2D fields based on fields with {@linkcode MetadataModel.FgProperties.FIELD_2D_VIEW_POSITION} extracted during {@linkcode Extract2DFields}.
 * @param fields2d
 * @returns Rearranged 2D fields.
 */
export function Reposition2DFields(fields2d: MetadataModel.I2DFields) {
	let repositionedFields = structuredClone(fields2d.fields)

	for (const fieldIndex of Object.keys(fields2d.reposition)) {
		const sourceIndex = Number(fieldIndex)

		let destinationIndex = -1
		for (let fIndex = 0; fIndex < repositionedFields.length; fIndex++) {
			if (repositionedFields[fIndex][MetadataModel.FgProperties.FIELD_GROUP_KEY] === fields2d.reposition[sourceIndex][MetadataModel.Field2dPositionProperties.FIELD_GROUP_KEY]) {
				if (typeof fields2d.reposition[sourceIndex][MetadataModel.Field2dPositionProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] === 'number') {
					if (repositionedFields[fIndex][MetadataModel.FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] !== fields2d.reposition[sourceIndex][MetadataModel.Field2dPositionProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX]) {
						continue
					}
				}

				if (fields2d.reposition[sourceIndex][MetadataModel.Field2dPositionProperties.FIELD_POSITION_BEFORE]) {
					destinationIndex = fIndex
				} else {
					destinationIndex = fIndex + 1
				}
			}
		}

		if (destinationIndex >= 0) {
			repositionedFields = [...repositionedFields.slice(0, destinationIndex), structuredClone(repositionedFields[sourceIndex]), ...repositionedFields.slice(destinationIndex, sourceIndex), ...repositionedFields.slice(sourceIndex + 1)]
		}
	}

	return repositionedFields
}

File: internal/pkg/metadata_model/2d_fields.go
package metadatamodel

import (
	"errors"
	"fmt"
	"regexp"

	"github.com/barkimedes/go-deepcopy"
)

// Extracts fields that can be used for working with data in 2D array form like in excel or csv.
//
// Recommended order to Extract fields:
//
// - Extract2DFields.Extract
//
// - Extract2DFields.Reposition
//
// - Extract2DFields.RemoveSkipped
//
// - Extract2DFields.Fields
type Extract2DFields struct {
	fields               []any
	skipIfFgDisabled     bool
	skipIfDataExtraction bool
	removePrimaryKey     bool
	metadataModel        map[string]any
	repositionFields     RepositionFields
}

// Constructor for Extract2DFields
//
// Parameters:
//
//   - metadatamodel - A valid metadata-model of type object (not array). Expected to presented as if converted from JSON.
//
//   - skipIfFGDisabled - Do not include field group if property FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE($FG_VIEW_DISABLE) is true.
//
//   - skipIfDataExtraction - Do not include field group if property FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION($DATABASE_SKIP_DATA_EXTRACTION) is true.
//
//   - removePrimaryKey - Remove field if FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE or FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION even if FIELD_GROUP_PROP_FIELD_GROUP_IS_PRIMARY_KEY field property is true.
//
// returns error if argument metadatamodel is not of type map[string]any.
func NewExtract2DFields(metadatamodel any, skipIfFGDisabled bool, skipIfDataExtraction bool, removePrimaryKey bool) (*Extract2DFields, error) {
	n := new(Extract2DFields)
	if metadatamodelMap, ok := metadatamodel.(map[string]any); ok {
		n.metadataModel = metadatamodelMap
	} else {
		return nil, errors.New("argument metadatamodelMap is not of type map[string]any")
	}
	n.skipIfFgDisabled = skipIfFGDisabled
	n.skipIfDataExtraction = skipIfDataExtraction
	n.removePrimaryKey = removePrimaryKey
	return n, nil
}

// Return information about fields that need to be repositioned based on FIELD_GROUP_PROP_FIELD_2D_VIEW_POSITION when Extract2DFields.Extract is called.
func (n *Extract2DFields) RepositionFields() RepositionFields {
	return n.repositionFields
}

// Return Extract2DFields.fields in its current state.
func (n *Extract2DFields) Fields() []any {
	return n.fields
}

// Extracts fields found in  Extract2DFields.metadataModel and places them in Extract2DFields.fields.
func (n *Extract2DFields) Extract() error {
	return n.extract(n.metadataModel, false, false)
}

// Recursive function that extracts fields found in mmGroup and places them in Extract2DFields.fields.
//
// Will not remove any fields if FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION or FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE is true.
//
// Parameters:
//
// - mmGroup - Current metadata model Group to extract fields from.
//
// - mmGroupSkipDataExtraction - Add FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION property to child fields and groups of mmGroup if true.
//
// - mmGroupViewDisable - Add FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE property to child fields and groups of mmGroup if true.
//
// mm - Alias for metadata model.
//
// fg - Alias for field group.
func (n *Extract2DFields) extract(mmGroup any, mmGroupSkipDataExtraction bool, mmGroupViewDisable bool) error {
	mmGroupMap, err := GetFieldGroupMap(mmGroup)
	if err != nil {
		return err
	}

	mmGroupFields, err := GetGroupFields(mmGroupMap)
	if err != nil {
		return err
	}

	mmGroupReadOrderOfFields, err := GetGroupReadOrderOfFields(mmGroup)
	if err != nil {
		return err
	}

	for _, fgKeySuffix := range mmGroupReadOrderOfFields {
		fgKeySuffixString, err := GetValueAsString(fgKeySuffix)
		if err != nil {
			return err
		}

		fgMap, err := GetFieldGroupMap(mmGroupFields[fgKeySuffixString])
		if err != nil {
			return err
		}

		if fgGroupFields, err := GetGroupFields(fgMap); err == nil {
			if fgGroupReadOrderOfFields, err := GetGroupReadOrderOfFields(fgMap); err == nil {
				skipDataExtraction := mmGroupSkipDataExtraction
				if value, ok := fgMap[FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION].(bool); ok {
					skipDataExtraction = value
				}
				viewDisable := mmGroupViewDisable
				if value, ok := fgMap[FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE].(bool); ok {
					viewDisable = value
				}

				if value, ok := fgMap[FIELD_GROUP_PROP_GROUP_EXTRACT_AS_SINGLE_FIELD].(bool); ok && value {
					var newField map[string]any = fgMap
					if value, err := deepcopy.Anything(fgMap); err == nil {
						if valueMap, ok := value.(map[string]any); ok {
							newField = valueMap
						}
					}

					n.fields = append(n.fields, newField)
					continue
				}

				if fgViewMaxNoOfValuesInSeparateColumns := FgGet2DConversion(fgMap); fgViewMaxNoOfValuesInSeparateColumns > 1 {
					for columnIndex := 0; columnIndex < fgViewMaxNoOfValuesInSeparateColumns; columnIndex++ {
						for _, gfKey := range fgGroupReadOrderOfFields {
							gfKeyString, err := GetValueAsString(gfKey)
							if err != nil {
								return err
							}

							gfMap, err := GetFieldGroupMap(fgGroupFields[gfKeyString])
							if err != nil {
								return err
							}

							var newField map[string]any = gfMap
							if value, err := deepcopy.Anything(gfMap); err == nil {
								if valueMap, ok := value.(map[string]any); ok {
									newField = valueMap
								}
							}

							n.updateSeparateColumnsField(newField, mmGroupSkipDataExtraction, mmGroupViewDisable, columnIndex)

							n.fields = append(n.fields, newField)
						}
					}
					continue
				}

				n.extract(fgMap, skipDataExtraction, viewDisable)
				continue
			}
		}

		if fgViewMaxNoOfValuesInSeparateColumns := FgGet2DConversion(fgMap); fgViewMaxNoOfValuesInSeparateColumns > 1 {
			for columnIndex := 0; columnIndex < fgViewMaxNoOfValuesInSeparateColumns; columnIndex++ {
				var newField map[string]any = fgMap
				if value, err := deepcopy.Anything(newField); err == nil {
					if valueMap, ok := value.(map[string]any); ok {
						newField = valueMap
					}
				}

				n.updateSeparateColumnsField(newField, mmGroupSkipDataExtraction, mmGroupViewDisable, columnIndex)

				n.fields = append(n.fields, newField)
			}

			continue
		}

		var newField map[string]any = fgMap
		if value, err := deepcopy.Anything(fgMap); err == nil {
			if valueMap, ok := value.(map[string]any); ok {
				newField = valueMap
			}
		}

		if value := Get2DFieldViewPosition(newField); value != nil {
			n.repositionFields[len(n.fields)] = value
		}

		if mmGroupSkipDataExtraction {
			newField[FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION] = true
		}

		if mmGroupViewDisable {
			newField[FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE] = true
		}

		n.fields = append(n.fields, newField)
	}

	return nil
}

func (n *Extract2DFields) updateSeparateColumnsField(field map[string]any, mmGroupSkipDataExtraction bool, mmGroupViewDisable bool, columnIndex int) {
	if mmGroupSkipDataExtraction {
		field[FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION] = true
	}

	if mmGroupViewDisable {
		field[FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE] = true
	}

	if fgHeaderFormat, ok := field[FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT].(string); ok && len(fgHeaderFormat) > 0 {
		re := regexp.MustCompile(fgHeaderFormat)
		field[FIELD_GROUP_PROP_FIELD_GROUP_NAME] = string(re.ReplaceAll([]byte(ARRAY_PATH_PLACEHOLDER), []byte(fmt.Sprintf("%d", columnIndex+1))))
	} else {
		field[FIELD_GROUP_PROP_FIELD_GROUP_NAME] = fmt.Sprintf("%s %v", GetFieldGroupName(field, "#unnamed"), columnIndex+1)
	}
}

// Calls Reposition2DFields and updates Extract2DFields.fields.
//
// Returns an error if Reposition2DFields returns an error.
func (n *Extract2DFields) Reposition() error {
	if newFields, err := Reposition2DFields(Field2DsWithReposition{Fields: n.fields, Reposition: n.repositionFields}); err != nil {
		return err
	} else {
		n.fields = newFields
	}

	return nil
}

// Returns result of calling Reposition2DFields.
func (n *Extract2DFields) FieldsRepositioned() ([]any, error) {
	return Reposition2DFields(Field2DsWithReposition{Fields: n.fields, Reposition: n.repositionFields})
}

// Calls RemoveSkipped2DFields and updates Extract2DFields.fields.
//
// Returns an error if RemoveSkipped2DFields returns an error.
func (n *Extract2DFields) RemoveSkipped() error {
	if newFields, err := RemoveSkipped2DFields(n.fields, n.skipIfFgDisabled, n.skipIfDataExtraction, n.removePrimaryKey); err != nil {
		return err
	} else {
		n.fields = newFields
	}

	return nil
}

func (n *Extract2DFields) FieldsWithSkippedRemoved() ([]any, error) {
	return RemoveSkipped2DFields(n.fields, n.skipIfFgDisabled, n.skipIfDataExtraction, n.removePrimaryKey)
}

// Reorder columns of each row in data to match order in targetFields.
type Reorder2DFields struct {
	sourceToTargetReadOrderOfFields []int
	targetToSourceReadOrderOfFields []int
}

// Constructor for Reorder2DFields.
//
// Initializes Reorder2DFields.sourceToTargetReadOrderOfFields and Reorder2DFields.targetToSourceReadOrderOfFields.
//
// Parameters:
//
//   - sourceFields - Current order of columns in each row of data.
//
//   - skipIfFGDisabled - Target order of columns in each row of data.
//
// returns error if sourceFields and targetFields are not valid.
func NewReorder2DFields(sourceFields []any, targetFields []any) (*Reorder2DFields, error) {
	n := new(Reorder2DFields)
	if len(sourceFields) != len(targetFields) {
		return nil, errors.New("length of sourceFields and targetFields not equal")
	}

	n.sourceToTargetReadOrderOfFields = make([]int, 0)
	n.targetToSourceReadOrderOfFields = make([]int, len(sourceFields))

	for tfIndex, tfValue := range targetFields {
		sourceIndex := -1

		tfValueMap, err := GetFieldGroupMap(tfValue)
		if err != nil {
			return nil, err
		}

		for sfIndex, sfValue := range sourceFields {
			sfValueMap, err := GetFieldGroupMap(sfValue)
			if err != nil {
				return nil, err
			}

			if sfValueMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY] == tfValueMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY] {
				if value := Get2DFieldViewPosition(sfValueMap); value != nil {
					if value.FViewValuesInSeparateColumnsHeaderIndex != nil {
						if tfValueMap[FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] != *value.FViewValuesInSeparateColumnsHeaderIndex {
							continue
						}
					}
				}
				n.sourceToTargetReadOrderOfFields = append(n.sourceToTargetReadOrderOfFields, sfIndex)
				sourceIndex = sfIndex
				break
			}
		}

		if sourceIndex < 0 {
			return nil, errors.New("targetField not found in sourceField")
		}

		n.targetToSourceReadOrderOfFields[sourceIndex] = tfIndex
	}

	return n, nil
}

// Reorder columns in each row in data. Will modify order of columns in rows of original data so deep copy may needed if original order is required.
//
// returns an error if length of datum in data and Reorder2DFields.sourceToTargetReadOrderOfFields are not equal.
func (n *Reorder2DFields) Reorder(data [][]any) error {
	for dIndex, dValue := range data {
		if len(dValue) != len(n.sourceToTargetReadOrderOfFields) {
			return fmt.Errorf("length of datum at index %d and sourceToTargetReadOrderOfFields not equal", dIndex)
		}

		targetDatum := make([]any, len(n.sourceToTargetReadOrderOfFields))
		for stIndex, stValue := range n.sourceToTargetReadOrderOfFields {
			tDatumData := data[dIndex][stValue]
			if value, err := deepcopy.Anything(data[dIndex][stValue]); err == nil {
				tDatumData = value
			}
			targetDatum[stIndex] = tDatumData
		}

		data[dIndex] = targetDatum
	}

	return nil
}

// Remove 2D fields if FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE($FG_VIEW_DISABLE) or FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION($DATABASE_SKIP_DATA_EXTRACTION) is true.
//
// Returns updated fields and an error field is invalid.
func RemoveSkipped2DFields(fields []any, skipIfFgDisabled bool, skipIfDataExtraction bool, removePrimaryKey bool) ([]any, error) {
	newFields := []any{}

	for _, fValue := range fields {
		fValueMap, err := GetFieldGroupMap(fValue)
		if err != nil {
			return nil, err
		}

		if removePrimaryKey {
			if value, ok := fValueMap[FIELD_GROUP_PROP_FIELD_GROUP_IS_PRIMARY_KEY].(bool); ok && value {
				continue
			}
		}

		if skipIfFgDisabled {
			if value, ok := fValueMap[FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE].(bool); ok && value {
				continue
			}
		}

		if skipIfDataExtraction {
			if value, ok := fValueMap[FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION].(bool); ok && value {
				continue
			}
		}

		newField := fValue
		if value, err := deepcopy.Anything(fValue); err == nil {
			newField = value
		}
		newFields = append(newFields, newField)
	}

	return newFields, nil
}

// Rearrange 2D fields based on fields with FIELD_GROUP_PROP_FIELD_2D_VIEW_POSITION extracted during Extract2DFields.
//
// Returns error if deep copy of fields2D.Fields fails or repositionedFields is not valid.
func Reposition2DFields(fields2D Field2DsWithReposition) ([]any, error) {
	repositionedFields := fields2D.Fields
	if value, err := deepcopy.Anything(fields2D.Fields); err != nil {
		return nil, err
	} else {
		if valueSlice, ok := value.([]any); ok {
			repositionedFields = valueSlice
		}
	}

	for sourceIndex, rValue := range fields2D.Reposition {
		destinationIndex := -1

		for rfIndex, rfValue := range repositionedFields {
			rfValueMap, err := GetFieldGroupMap(rfValue)
			if err != nil {
				continue
			}

			if rfValueMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY] == rValue.FgKey {
				if rValue.FViewValuesInSeparateColumnsHeaderIndex != nil {
					if rfValueMap[FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX] != *rValue.FViewValuesInSeparateColumnsHeaderIndex {
						continue
					}
				}

				if rValue.FieldPositionBefore != nil && *rValue.FieldPositionBefore {
					destinationIndex = rfIndex
				} else {
					destinationIndex = rfIndex + 1
				}
			}
		}

		if destinationIndex >= 0 {
			newRepositionFields := repositionedFields[0:destinationIndex]
			newRepositionFields = append(newRepositionFields, repositionedFields[sourceIndex])
			newRepositionFields = append(newRepositionFields, repositionedFields[destinationIndex:sourceIndex]...)
			newRepositionFields = append(newRepositionFields, repositionedFields[sourceIndex+1:]...)
			repositionedFields = newRepositionFields
		}
	}

	return repositionedFields, nil
}
